pkg/client/ssa_client.go

```diff
@@ -0,0 +1,218 @@
+		config.determineUpdate = value
+	}
+}
+
+// EnsureLabels makes sure that the provided labels are applied to the object even if
+// the supplied object doesn't have them set.
+func EnsureLabels(labels map[string]string) SsaApplyObjectOption {
+	return func(config *ssaApplyObjectConfiguration) {
+		config.newLabels = labels
+	}
+}
+
+// Configure sets the owner reference and merges the labels. Other options modify the logic
+// of apply function and therefore need to be checked manually.
+func (c *ssaApplyObjectConfiguration) Configure(obj client.Object, s *runtime.Scheme) error {
+	if c.owner != nil {
+		if err := controllerutil.SetControllerReference(c.owner, obj, s); err != nil {
+			return err
+		}
+	}
+	MergeLabels(obj, c.newLabels)
+
+	return nil
+}
+
+// ApplyObject creates the object if is missing and if the owner object is provided, then it's set as a controller reference.
+// If the objects exists then when the spec content has changed (based on the content of the annotation in the original object) then it
+// is automatically updated. If it looks to be same then based on the value of forceUpdate param it updates the object or not.
+// The return boolean says if the object was either created or updated (`true`). If nothing changed (ie, the generation was not
+// incremented by the server), then it returns `false`.
+//
+// NOTE: the return boolean IS ALWAYS TRUE on success by default. This is much more efficient. If you truly need
+// to determine whether the object has been updated or not (which is not the case anywhere in the codebase but the tests),
+// you can use the DetermineUpdate() option.
+func (c *SsaApplyClient) ApplyObject(ctx context.Context, obj client.Object, options ...SsaApplyObjectOption) (bool, error) {
+	config := newSsaApplyObjectConfiguration(options...)
+	if err := config.Configure(obj, c.Client.Scheme()); err != nil {
+		return false, err
+	}
+
+	if err := prepareForSSA(obj, c.Client.Scheme()); err != nil {
+		return false, err
+	}
+
+	if config.skipIf != nil && config.skipIf(obj) {
+		return false, nil
+	}
+
+	updated := true
+
+	// NOTE: once the SSA migration is not needed anymore, read the orig conditionally only when config.determineUpdate is true.
+
+	// this cannot fail - we would have failed previously already because the scheme wouldn't have known the object's kind
+	origRo, _ := c.Client.Scheme().New(obj.GetObjectKind().GroupVersionKind())
+	orig := origRo.(client.Object)
+
+	if err := c.Client.Get(ctx, client.ObjectKeyFromObject(obj), orig); err != nil && !apierrors.IsNotFound(err) {
+		return false, err
+	}
+
+	// NOTE: remove this once this code is used for all "apply-style" functionality and has updated all the objects in the cluster.
+	if isSsaMigrationNeeded(orig, c.NonSSAFieldOwner) {
+		if err := c.MigrateToSSA(ctx, orig); err != nil {
+			return false, err
+		}
+	}
+
+	if err := c.Client.Patch(ctx, obj, client.Apply, client.FieldOwner(c.FieldOwner), client.ForceOwnership); err != nil {
```

Comment:
do I understand it correctly that in case of migration, we would need to do update and patch the resource immediately after each other?
Could the Patch call cause any conflicts? I see that it's not taking the resourceVersion of the updated object.

Reply:
yes, and it's intentional here, because that will remove any additional "cruft" from the resource that we applied to it using the old logic, like the `last-applied-configuration`.

Reply:
Patch will not cause a conflict. Resource version doesn't have to be set - that's the general case with patches actually - they're supposed to work without getting the object first and therefore without knowing its resource version.